File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int n;
    int fun(int i,int target,vector<int>&coins,vector<vector<int>>&dp){
        if(target==0) return 0;
        if(i==n-1){
            if(target%coins[i]==0){
                return target/coins[i];
            }
            else return 1e9;
        }
        if(dp[i][target]!=-1) return dp[i][target];
        int nottake=fun(i+1,target,coins,dp);
        int take=INT_MAX;
        if(target>=coins[i]){
            take=1+fun(i,target-coins[i],coins,dp);
        }
        return dp[i][target]=min(take,nottake);
    }
    int coinChange(vector<int>& coins, int amount) {
        n=coins.size();
        if(n==1){
            if(amount%coins[0]==0){
                return amount/coins[0];
            }
            else{
                return -1;
            }
        }
        vector<vector<int>>dp(n,vector<int>(amount+1,-1));
        int ans= fun(0,amount,coins,dp);
        return ans==1e9?-1:ans;
      

    }
};

File: 0322-coin-change/0322-coin-change.py
class Solution:
    def fun(self, i, amt, coins, dp):

        if amt == 0:
            return 0

        if i == len(coins) or amt < 0:
            return float('inf')

        if dp[i][amt] != -1:
            return dp[i][amt]

        take = 1 + self.fun(i, amt - coins[i], coins, dp)

        nottake = self.fun(i + 1, amt, coins, dp)

        dp[i][amt] = min(take, nottake)

        return dp[i][amt]

    def coinChange(self, coins: List[int], amount: int) -> int:

        n = len(coins)
      
        dp = [[-1] * (amount + 1) for _ in range(n)]

        ans = self.fun(0, amount, coins, dp)

        if ans == float('inf'):
            return -1

        return ans

